subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2031 @@
+	}
+
+	if (!is_char(buf, '=')) {
+		return -EINVAL;
+	}
+
+	if (is_char(buf, '?')) {
+		if (!is_char(buf, '\r')) {
+			return -EINVAL;
+		}
+
+		data = &ag->buffer[0];
+		*data = '(';
+		data++;
+		hf_indicators = ag->hf_indicators_of_ag;
+		len = (sizeof(hf_indicators) * 8) > HFP_HF_IND_MAX ? HFP_HF_IND_MAX
+								   : (sizeof(hf_indicators) * 8);
+		for (int i = 1; (i < len) && (hf_indicators != 0); i++) {
+			if (BIT(i) & hf_indicators) {
+				int length = snprintk(
+					data, (char *)&ag->buffer[HF_MAX_BUF_LEN - 1] - data - 3,
+					"%d", i);
+				data += length;
+				hf_indicators &= ~BIT(i);
+			}
+			if (hf_indicators != 0) {
+				*data = ',';
+				data++;
+			}
+		}
+		*data = ')';
+		data++;
+		*data = '\r';
+		data++;
+		*data = '\0';
+		data++;
+
+		err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BIND:%s\r\n", &ag->buffer[0]);
+		return err;
+	}
+
+	err = 0;
+	while (buf->len > 0) {
+		err = get_number(buf, &number);
+		if (err != 0) {
+			return err;
+		}
+
+		if (!is_char(buf, ',')) {
+			if (!is_char(buf, '\r')) {
+				return -EINVAL;
+			}
+		}
+
+		if (number < (sizeof(hf_indicators) * 8)) {
+			hf_indicators |= BIT(number);
+		}
+	}
+
+	ag->hf_indicators_of_hf = hf_indicators;
+
+	return 0;
+}
+
+static int bt_hfp_ag_cmee_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *cme_err)
+{
+	uint32_t number;
+	int err;
+
+	*cme_err = CME_ERROR_AG_FAILURE;
+
+	if (!is_char(buf, '=')) {
+		return -EINVAL;
+	}
+
+	err = get_number(buf, &number);
+	if (err != 0) {
+		return err;
+	}
+
+	if (!is_char(buf, '\r')) {
+		return -EINVAL;
+	}
+
+	if (number == 1) {
+		atomic_set_bit(ag->flags, BT_HFP_AG_CMEE_ENABLE);
+	} else if (number == 0) {
+		atomic_clear_bit(ag->flags, BT_HFP_AG_CMEE_ENABLE);
+	} else {
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int hfp_ag_update_indicator(struct bt_hfp_ag *ag, enum bt_hfp_ag_indicator index,
+				   uint8_t value, bt_hfp_ag_tx_cb_t cb, void *user_data)
+{
+	int err;
+	unsigned int key;
+	uint8_t old_value;
+
+	key = irq_lock();
+	old_value = ag->indicator_value[index];
+	if (value == old_value) {
+		LOG_ERR("Duplicate value setting, indicator %d, old %d -> new %d", index, old_value,
+			value);
+		irq_unlock(key);
+		return -EINVAL;
+	}
+
+	ag->indicator_value[index] = value;
+
+	irq_unlock(key);
+
+	LOG_DBG("indicator %d, old %d -> new %d", index, old_value, value);
+
+	err = hfp_ag_send_data(ag, cb, user_data, "\r\n+CIEV:%d,%d\r\n", (uint8_t)index + 1,
+				 value);
+	if (err) {
+		key = irq_lock();
+		ag->indicator_value[index] = old_value;
+		irq_unlock(key);
+	}
+
+	return err;
+}
+
+static void hfp_ag_close_sco(struct bt_hfp_ag *ag)
+{
+	LOG_DBG("");
+	if (NULL != ag->sco_chan.sco) {
+		bt_conn_disconnect(ag->sco_chan.sco, BT_HCI_ERR_LOCALHOST_TERM_CONN);
+		ag->sco_chan.sco = NULL;
+	}
+}
+
+static void bt_hfp_ag_reject_cb(struct bt_hfp_ag *ag, void *user_data)
+{
+	hfp_ag_close_sco(ag);
+	bt_hfp_ag_set_call_state(ag, BT_HFP_CALL_TERMINATE);
+
+	if (bt_ag && (bt_ag->reject)) {
```

Comment:
The `()` around `bt_ag->reject` is unnecessary

Reply:
updated